binary-search-tree/node-js/inventory-management.js

```diff
@@ -19,15 +19,43 @@ class BinarySearchTree {
     }
 
     search(id) {
-        console.error("you must implement search");
+        // console.error("you must implement search");
+        return this.exampleSearch(id);
+    }
+    exampleSearch(id) {
+        if (this.root.val.id = id) {
+            return this.root;
+        } else {
+            this.checkId(id, this.root);
+        }
+    }
+    checkId(id, node) {
+        if (node === null) {
+            return null;
+        } else {
+            if (id < node.val.id) {
+                if (node.left === id) {
+                    return node.left;
+                } else {
+                    this.checkId(id, node.left)
+                }
+            } else if (id > node.val.id) {
+                if (node.right === id) {
+                    return node.right
+                } else {
+                    this.checkId(id, node.right)
+                }
+            }
+        }
     }
```

Comment:
1. The general approach is good. I can see that you're using the properties of a binary search tree in implementing this method e.g. you only check the left subtree if the search value is less than the current node value.
2. When you recursively call `checkId()`, you need to return the result of that call. So instead of just calling `this.checkId(id, node.left)` or `this.checkId(id, node.right)`, you should return the result of the call, like `return this.checkId(id, node.left)` or `return this.checkId(id, node.right)`.
3. When checking if the left or right child is equal to `id`, you are using the `===` operator, which checks for strict equality of both value and type. However, `node.left` and `node.right` are instances of the `TreeNode` class, not integers like `id`. So the comparison will always be `false` and the method will continue recursively checking down the tree indefinitely. 

   To fix this, you would check the `id` property of the `val` property of the child node, like you did in the `exampleSearch` method: `if (node.left.val.id == id)` and `if (node.right.val.id == id)`.
4. Related to (3). You're checking `node.left` and `node.right` against the search `id`. You're duplicating work there. 

   To reduce the duplication,  you can update `checkId` to check the current node against the search `id` before you check the left and right child nodes.  Depending on the branch you can immediately recurse on the left or right child nodes, knowing that the recursive call will handle the check. This means your recursive `checkId` function has 4 cases, an empty node, a node that matches the search value, a node that is greater than the search value, a node that is less than the search value.

Reply:
I think you've got it! It's just that `search` should return the node value instead of the node.

Reply:
I think you've got it! It's just that search should return the node value instead of the node.